### DISCUSSION TITLE
Multiple OSM files/regions?

### DISCUSSION THREAD

**Discussion by MTRNord** [motis-project/motis#513] [Q&A] [answered]
Hi I was wondering: What is the correct way to have multiple regions in motis?

The config I tried is:

```
# Available moduels: ["adr", "gbfs", "intermodal", "nigiri", "osr", "osrm", "parking", "ppr", "tiles"]
modules=intermodal
modules=adr
modules=tiles
modules=osrm
modules=ppr
modules=parking
modules=nigiri
#modules=gbfs

intermodal.router=nigiri

[import]
paths=schedule-de:/input/schedule/latest.zip
paths=osm-de:/input/osm-de.pbf
paths=osm-dk:/input/osm-dk.pbf
data_dir=/data

[nigiri]
first_day=TODAY
num_days=31
default_timezone=Europe/Berlin
match_duplicates=true
gtfsrt=de|https://realtime.gtfs.de/realtime-free.pb

[osrm]
profiles=/motis/osrm-profiles/car.lua
profiles=/motis/osrm-profiles/bike.lua
profiles=/motis/osrm-profiles/bus.lua
#profiles=/motis/osrm-profiles/foot.lua
```

but that does not seem to generate any tiles for denmark.

I also tried merging it into a single file but that leads to ppr failing with bad_alloc after the serializing step.

**Comment by felixguendling** [comment, marked as answer]
I usually use the osmium command line tools and create an extract from the next bigger region containing all regions I need (in this case it's probably the Europe extract from Geofabrik). I create a very rough polygon using [geojson.io](https://geojson.io) and give this GeoJSON + the Europe OSM file to the osmium command line tool.


If you do not really need the `parking` module and have a SSD (no HDD), I would recommend to try it with this configuration and use the full European OSM data:

```ini
modules=intermodal
modules=adr
modules=tiles
modules=osr
modules=nigiri

[import]
paths=schedule-de:/input/schedule/latest.zip
paths=osm-de:/input/europe-latest.osm.pbf
data_dir=/data

[nigiri]
num_days=31
default_timezone=Europe/Berlin
match_duplicates=true
gtfsrt=de|https://realtime.gtfs.de/realtime-free.pb
```

I replaced `ppr` and `osrm` with our new "in-house" routing `osr` which requires way less memory.

We plan to add sharing support (GBFS) as well as parking support to `osr`, so `osr` will be able to fulfill all use cases.

**Reply by felixguendling** [reply]
Your `tiles` configuration does not contain coastlines. Since your area is close to the water, I would add coastlines.

```ini
modules=intermodal
modules=adr
modules=tiles
modules=osr
modules=nigiri

tiles.import.use_coastline=true

[import]
paths=schedule-de:/input/schedule/latest.zip
paths=osm-de:/input/europe-latest.osm.pbf
paths=coastline:/input/land-polygons-complete-4326.zip
data_dir=/data

[nigiri]
num_days=31
default_timezone=Europe/Berlin
match_duplicates=true
gtfsrt=de|https://realtime.gtfs.de/realtime-free.pb
```

Warning: Processing the coastlines takes some time and more memory.

**Reply by felixguendling** [reply]
Another thing: you're currently using the default background map style configured in the Dockerfile I guess (`tiles.profile=/motis/tiles-profiles/background.lua`).

If you want to have a more colorful style as shown [in the v0.11.30 release](https://github.com/motis-project/motis/releases/tag/v0.11.30), you would need to switch to `tiles.profile=/motis/tiles-profiles/profile.lua`.

```ini
modules=intermodal
modules=adr
modules=tiles
modules=osr
modules=nigiri

[import]
paths=schedule-de:/input/schedule/latest.zip
paths=osm-de:/input/europe-latest.osm.pbf
paths=coastline:/input/land-polygons-complete-4326.zip
data_dir=/data

[nigiri]
num_days=31
default_timezone=Europe/Berlin
match_duplicates=true
gtfsrt=de|https://realtime.gtfs.de/realtime-free.pb

[tiles]
import.use_coastline=true
profile=/motis/tiles-profiles/profile.lua
```

You can try the colorful style here: https://europe.motis-project.de/?style=color

**Reply by MTRNord** [reply]
Thanks a lot for all the options! :) This really helps. Also thanks for everyone working on this :) Seems like it might be a replacement for OTP for us :D

**Reply by felixguendling** [reply]
Happy to hear that! Feel free to ask more questions or file bug reports if you encounter a problem.

**Reply by MTRNord** [reply]
I do actually have a followup question which may be a dedicated bug in the end :)

I am seeing this weird thing happening:

![grafik](https://github.com/motis-project/motis/assets/1374914/d9dc3b1c-9d21-4c94-8812-25cbaa742f00)


any ideas what may cause this? :thinking: the gtfs data is just https://gtfs.de/de/feeds/ in the full version. So I _assume_ its correct. But clearly something is breaking :)

**Reply by MTRNord** [reply]
![grafik](https://github.com/motis-project/motis/assets/1374914/70015e91-44fa-497f-a79e-59a6c5f5d9bf)

The gtfs-rt data is also very confused :D

**Reply by felixguendling** [reply]
I would always recommend to use the official data source for the German timetable. `gtfs.de` and its real-time data is a proprietary dataset which isn't realiable in my experience.

This is the correct source: https://www.opendata-oepnv.de/ht/de/willkommen
(*) Edit: This is also the data we use for europe.motis-project.de or for transitous.org

There's no real-time data but it has been announced that real-time data will be available this year. So let's see :)

**Reply by felixguendling** [reply]
I tried to replicate the problem you had but for the exact same query, the problem was not reproducable. Did you by chance download the GTFS-RT protobuf file at the time your problem occurred?

**Reply by felixguendling** [reply]
> ![grafik](https://private-user-images.githubusercontent.com/1374914/335878839-70015e91-44fa-497f-a79e-59a6c5f5d9bf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTczNTQ2OTIsIm5iZiI6MTcxNzM1NDM5MiwicGF0aCI6Ii8xMzc0OTE0LzMzNTg3ODgzOS03MDAxNWU5MS00NGZhLTQ5N2YtYTc5ZS01OWE2YzVmNWQ5YmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDJUMTg1MzEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjQ2MTdiZGNiMmMzYmUyMzIyMGYzNWQxNGI4YTQ3NDY1NGEyODY2Mzk2ZTE0N2Y2NTU3OWQ0YmYzZTk3MGNkYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.mbtqdcAoy2JnS6Kvwmj5SH4MG0T16EGbKuONj-ya46w)
> 
> The gtfs-rt data is also very confused :D

And can you please next time also copy the URL? This contains important information such as the GTFS `trip_id`. With this ID it's possible to `grep` (or better `ripgrep`) through the static GTFS timetable data and the GTFS-RT (converted to JSON). Then it becomes way easier to diagnose what's actually going on. In my replicated setup I cannot find the crazy trip you have in your data.

**Reply by felixguendling** [reply]
This bus FM is in the area of the RMV transit agancy. There's no real-time data available for this trip. So I would really be interested, how to reproduce this :-)

**Reply by MTRNord** [reply]
yeah I am also still waiting 2 years now on rt data from the local company :D (technicians 2 years ago told me it will happen and it did not so far :))

**Reply by MTRNord** [reply]
> I tried to replicate the problem you had but for the exact same query, the problem was not reproducable. Did you by chance download the GTFS-RT protobuf file at the time your problem occurred?

oops I totally missed these messages....

So the rt data is synced at all time here using the gtfs.de. It happens still for me even after rebuilding. Looking at the station names I am wondering if there are some IDs overlapping maybe since the names of the station in the south are not part of the routes at all. So it can't be like a name mismatch with 2 stations using the same name.

> And can you please next time also copy the URL? 

By the url you mean the url in the browser correct?

**Reply by felixguendling** [reply]
> By the url you mean the url in the browser correct?

Yes.

To reproduce I need the exact GTFS-RT protobuf file + the trip ID (which is contained in the URL).

I tried again to reproduce the issue but on my setup (with the gtfs.de datasets) the journeys look good.

**Reply by felixguendling** [reply]
> it can't be like a name mismatch with 2 stations using the same name

Station names are irrelevant for matching trip updates. Only station IDs are used. As long as those are unique, everything should be fine.

**Reply by felixguendling** [reply]
Are you using an up-to-date GTFS-static dataset? It's possible that things like this happen if the GTFS-static dataset is not up-to-date and IDs change. But then, any journey planner (not just MOTIS) would have issues.

**Reply by felixguendling** [reply]
I set up the MOTIS instance as a server with automatic restart 1x per day (3:30 am): https://gtfsde.motis-project.de/

**Reply by MTRNord** [reply]
> Are you using an up-to-date GTFS-static dataset? It's possible that things like this happen if the GTFS-static dataset is not up-to-date and IDs change. But then, any journey planner (not just MOTIS) would have issues.

hm its a few days old when I started to look into motis the first time. I hoped that in the 3 days timespan the gtfs file would have stayed recent enough but this certainly may be the root cause here.

**Reply by MTRNord** [reply]
I will take a look at this on the weekend when I got some time for it and try to get your the necessary data. I will first update the gtfs-static dataset to see if thats the cause.

**Reply by felixguendling** [reply]
I had now the same problem and it really seems to be that the feed does not use stable IDs. Almost all feeds try to use IDs that are somehow stable across different versions of the schedule timetable (GTFS static). This feed however uses just auto-incrementing IDs with no specific order. So if their server (gtfs.de) changes the GTFS-static file you *have to* update in the same second, otherwise your real-time timetable will be completely messed up. Other GTFS-static/GTFS-RT endpoints are not that problematic as most try to use a stable referencing scheme for trips and stops.
